src/btree/bt_handle.c

```diff
@@ -15,6 +15,39 @@ static int __btree_preload(WT_SESSION_IMPL *);
 static int __btree_tree_open_empty(WT_SESSION_IMPL *, bool);
 
 /*
+ * __btree_initialize --
+ *	Initialize the WT_BTREE structure.
+ */
+static void
+__btree_initialize(WT_BTREE *btree, bool closing)
+{
+	uint32_t mask;
+
+	/*
+	 * This function exists as a place to talk about how the WT_BTREE
+	 * structure is initialized (or re-initialized, when the object is
+	 * re-opened). The upper-level handle code currently sets/clears
+	 * information in the WT_BTREE structure, specifically flags, cache
+	 * and eviction information. All of that information must persist
+	 * after the WT_BTREE object is closed, and some of it has already
+	 * been initialized when the object is opened/re-opened.
+	 */
+	if (closing) {
+		/*
+		 * Closing: clear everything except cache/eviction information
+		 * and one LSM flag.
+		 */
+		memset(btree, 0, WT_BTREE_CLEAR_SIZE(btree));
+		F_CLR(btree, ~(WT_BTREE_LSM_PRIMARY | WT_BTREE_NO_EVICTION));
```

Comment:
Failing to preserve the flag makes `test_cursor06` fail:
```
test_cursor06.test_cursor06.test_reconfigure_overwrite(lsm-S): ,
file:reconfigure-000002.lsm, WT_CONNECTION.close:
WT_BTREE.bytes_dirty_leaf underflow: decrementing 11608
```
The problem is drop closes the btree handle, but the bytes aren't discarded until later. Here's the stack from drop closing the handle:
```
#0  __btree_initialize (btree=0x807440380, closing=true)
    at src/btree/bt_handle.c:43
#1  0x0000000803aa83ed in __wt_btree_close (session=0x8071618a0)
    at src/btree/bt_handle.c:227
#2  0x0000000803afe698 in __wt_conn_btree_sync_and_close (session=0x8071618a0, 
    final=false, force=true) at src/conn/conn_dhandle.c:194
#3  0x0000000803aff7e4 in __wt_conn_dhandle_close_all (session=0x8071618a0, 
    uri=0x8074230a0 "file:reconfigure_cgroup1-000002.lsm", force=true)
    at src/conn/conn_dhandle.c:508
#4  0x0000000803ba86c9 in __drop_file (session=0x8071618a0, 
    uri=0x8074230a0 "file:reconfigure_cgroup1-000002.lsm", force=true, 
    cfg=0x7fffffffcb80) at src/schema/schema_drop.c:33
#5  0x0000000803ba8148 in __wt_schema_drop (session=0x8071618a0, 
    uri=0x8074230a0 "file:reconfigure_cgroup1-000002.lsm", cfg=0x7fffffffcb80)
    at src/schema/schema_drop.c:181
#6  0x0000000803b6bf9a in __wt_lsm_tree_drop (session=0x8071618a0, 
    name=0x8070e5e20 "lsm:reconfigure_cgroup1", cfg=0x7fffffffcb80)
    at src/lsm/lsm_tree.c:919
#7  0x0000000803ba81ee in __wt_schema_drop (session=0x8071618a0, 
    uri=0x8070e5e20 "lsm:reconfigure_cgroup1", cfg=0x7fffffffcb80)
    at src/schema/schema_drop.c:185
#8  0x0000000803ba8a84 in __drop_table (session=0x8071618a0, 
    uri=0x806e2686c "table:reconfigure", cfg=0x7fffffffcb80)
    at src/schema/schema_drop.c:129
#9  0x0000000803ba823d in __wt_schema_drop (session=0x8071618a0, 
    uri=0x806e2686c "table:reconfigure", cfg=0x7fffffffcb80)
    at src/schema/schema_drop.c:187
#10 0x0000000803bbc339 in __session_drop (wt_session=0x8071618a0, 
    uri=0x806e2686c "table:reconfigure", config=0x806dc6984 "force")
    at src/session/session_api.c:856
```
and then we eventually close the connection:
```
#6  0x0000000803aa6027 in __wt_cache_page_byte_dirty_decr (
    session=0x807027000, page=0x80780e1a0, size=11608) at btree.i:295
#7  0x0000000803aa5de2 in __wt_cache_dirty_decr (session=0x807027000, 
    page=0x80780e1a0) at btree.i:388
#8  0x0000000803aa5032 in __wt_page_modify_clear (session=0x807027000, 
    page=0x80780e1a0) at btree.i:594
#9  0x0000000803aa4ca0 in __wt_page_out (session=0x807027000, 
    pagep=0x808011100) at src/btree/bt_discard.c:80
#10 0x0000000803aa4c43 in __wt_ref_out (session=0x807027000, ref=0x808011100)
    at src/btree/bt_discard.c:59
#11 0x0000000803b3fa88 in __wt_evict_file (session=0x807027000, 
    syncop=WT_SYNC_DISCARD) at src/evict/evict_file.c:91
#12 0x0000000803ad1868 in __wt_cache_op (session=0x807027000, 
    op=WT_SYNC_DISCARD) at src/btree/bt_sync.c:319
#13 0x0000000803bdcf87 in __wt_checkpoint_close (session=0x807027000, 
    final=true) at src/txn/txn_ckpt.c:1641
#14 0x0000000803afee8a in __wt_conn_btree_sync_and_close (session=0x807027000, 
    final=true, force=false) at src/conn/conn_dhandle.c:191
#15 0x0000000803b00142 in __wt_conn_dhandle_discard_single (
    session=0x807027000, final=true, force=false)
    at src/conn/conn_dhandle.c:575
#16 0x0000000803b00721 in __wt_conn_dhandle_discard (session=0x807027000)
    at src/conn/conn_dhandle.c:642
#17 0x0000000803b053cd in __wt_connection_close (conn=0x8023dc000)
    at src/conn/conn_open.c:131
#18 0x0000000803af3d6d in __conn_close (wt_conn=0x8023dc000, config=0x0)
    at src/conn/conn_api.c:1089
```
and `__wt_cache_page_byte_dirty_decr` is looking for that flag in the btree handle.